### PR TITLE
Qt5 reset signals after non-interactive plotting

### PR DESCRIPTION
**Pull request by joelfrederico**
## PR Summary

Non-interactive plots change the signal handler after they have been closed. They should reset it instead

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant

### CODE REVIEW

**Comment by joelfrederico** [comment]
I'm not sure how to test this without actually opening a window. I need to fire a non-interactive plot in Qt5. I don't need it to actually draw, but I need to run `show()`. I think some other tests open windows, so perhaps this is okay?

File: lib/matplotlib/tests/test_backend_qt.py [hunk 1]
```diff
@@ -114,6 +114,42 @@ def test_fig_close(backend):
     assert init_figs == Gcf.figs
 
 
+@pytest.mark.backend('Qt5Agg')
+def test_fig_close_ioff(qt_module):
+    # Turn off interactive mode
+    plt.ioff()
```
Comment (anntzer) [inline lib/matplotlib/tests/test_backend_qt.py]: do you need this line for the tests to pass?

Comment (joelfrederico) [reply lib/matplotlib/tests/test_backend_qt.py]: I think so, yes. Otherwise plt.show() isn't a blocking call, which I think means that an entirely different branch of code is executed. The test would pass, but it wouldn't be testing the fix I made. But I'm working off of memory, I can double-check later.

Comment (anntzer) [reply lib/matplotlib/tests/test_backend_qt.py]: Not sure this really matters, but OTOH it probably doesn't hurt, so heh.

Comment (joelfrederico) [reply lib/matplotlib/tests/test_backend_qt.py]: Oh that's true, it might be a no-op...

File: lib/matplotlib/tests/test_backend_qt.py [hunk 2]
```diff
@@ -114,6 +114,42 @@ def test_fig_close(backend):
...
+    original_signal = signal.getsignal(signal.SIGINT)
+
+    # Use our own SIGINT handler to be 100% sure this is working
+    def CustomHandler(signum, frame):
+        pass
+
+    signal.signal(signal.SIGINT, CustomHandler)
+
+    # Show window- fires callback and exits
+    plt.show()
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_backend_qt.py]: Instead `show` it may be more reliable to start the qt event loop.

**Comment by anntzer** [comment]
@tacaswell codecov seems happy (https://codecov.io/gh/matplotlib/matplotlib/compare/22649b6e8d72bbf142b5ec5824b1752b6ab0f557...745a4e9e20bdf4f76b83796aa4b1b8fe47086276/src/lib/matplotlib/tests/test_backend_qt.py?before=lib/matplotlib/tests/test_backend_qt.py#L115).  Not sure about ioff() though.

File: lib/matplotlib/tests/test_backend_qt.py (hunk 2 above)
Comment (joelfrederico) [reply lib/matplotlib/tests/test_backend_qt.py]: Did this in the new commit

**Comment by joelfrederico** [comment]
Okay, now I'm done! Sorry @tacaswell, @anntzer for the multiple commits. This should be pretty thorough now, testing the signals during and after the event loop. I don't actually fire signals because Python doesn't let you fire SIGINT on Windows (see https://docs.python.org/3/library/os.html#os.kill). You can do it with win32api, but I don't want to create a testing dependency on pywin32.

File: lib/matplotlib/tests/test_backend_qt.py [hunk 3]
```diff
@@ -114,6 +114,53 @@ def test_fig_close(backend):
     assert init_figs == Gcf.figs
 
 
+@pytest.mark.backend('Qt5Agg')
+def test_fig_close_ioff(qt_module):
```
Comment (anntzer) [inline lib/matplotlib/tests/test_backend_qt.py]: Rename the test perhaps?  Right now the name seems unrelated.

Comment (joelfrederico) [reply lib/matplotlib/tests/test_backend_qt.py]: Good call, renamed.

**Comment by tacaswell** [comment]
Thanks @joelfrederico !
